File: DataStructure/U1/Lecture/topoSort.py
import intReader

def TopoSort():
    print ("Topologischen Sortieren.")
    # Einlesen
    input = intReader.readInt()
    n = next(input)

    # Inititalisieren array:
    Knotenliste = [None] + [ Knoten(i) for i in range(1,n+1) ]

    
    # Einlesen Kanten
    try:
        while True:
            e = Knotenliste[next(input)]
            f = Knotenliste[next(input)]
            x = Kante(e,f)
            x.next = e.ersterNachfolger
            e.ersterNachfolger = x
            f.anzVorgänger += 1
    except StopIteration:
        pass
        
    # Vorbereiten:
    freieKnoten = [None]*n
    # explizit programmierter Stapel;
    # könnte auch mit append() (=push) und pop() geschrieben werden.    
    AnzfreieKnoten = 0;
    for i in range(1,n+1):
        e = Knotenliste[i]
        if e.anzVorgänger==0:
            freieKnoten[AnzfreieKnoten]=e
            AnzfreieKnoten += 1

    # Sortieren:
    print("");
    for i in range(1,n+1):
        if AnzfreieKnoten == 0:
            print("Die Eingabe enthält einen Kreis.")
            return
        # Wähle einen Knoten ohne Vorgänger
        AnzfreieKnoten -= 1
        x = freieKnoten[AnzfreieKnoten]
        print(x)
        # Entferne ausgehende Kanten:
        z = x.ersterNachfolger
        while z != None:
            z.v.anzVorgänger -= 1
            if z.v.anzVorgänger == 0:
                freieKnoten[AnzfreieKnoten]=z.v
                AnzfreieKnoten += 1
            z = z.next

class Knoten:
    def __init__(self, i):
        self.Name = i
        self.anzVorgänger = 0
        self.ersterNachfolger = None
    def __str__(self):
        return str(self.Name)

class Kante:
    def __init__(self, u, v):
        self.u = u
        self.v = v

TopoSort()        

File: DataStructure/U1/Lecture/TopoSort.java
import java.io.*;

class TopoSort
{ 
    int n;
    Knoten [] Knotenliste; // Liste aller Knoten 1,...,n
    Knoten [] freieKnoten; // Menge Q der Knoten ohne Vorgänger
    int AnzfreieKnoten;    // (als Stapel gespeichert)

    static IntReader input;

public static void main (String[] args) throws IOException
    {
	input = new IntReader(args);
	TopoSort x = new TopoSort();
    }

TopoSort() throws IOException
{
    System.out.println("Topologischen Sortieren.");
    // Einlesen
    n = input.readInt();
        
    // Inititalisieren array:
    Knotenliste = new Knoten [n+1];
    for (int i=1; i<=n; i++) Knotenliste[i] = new Knoten(i);

    // Einlesen Kanten
    try	{ 
	for (;;) {
	    Knoten e = Knotenliste[input.readInt()];
	    Knoten f = Knotenliste[input.readInt()];
	    Kante x = new Kante(e,f);
	    x.next = e.ersterNachfolger;
	    e.ersterNachfolger = x;
	    f.anzVorgänger++;
    }   }
    catch(EOFException e) {}
	
    // Vorbereiten:
    freieKnoten = new Knoten [n];
    AnzfreieKnoten = 0;
    for (int i=1; i<=n; i++) {
	Knoten e = Knotenliste[i];
	if (e.anzVorgänger==0)
	    freieKnoten[AnzfreieKnoten++]=e;
    }
    // Sortieren:
    System.out.println("");
 sort:
    for (int i=1; i<=n; i++) {
	if (AnzfreieKnoten == 0) {
	    System.out.println("Die Eingabe enthält einen Kreis."); 
	    break sort;
	}
	// Wähle einen Knoten ohne Vorgänger
	Knoten x = freieKnoten[--AnzfreieKnoten];
	System.out.println(x);
	// Entferne ausgehende Kanten:
	Kante z = x.ersterNachfolger;
	while (z != null) {
	    z.v.anzVorgänger--;
	    if (z.v.anzVorgänger == 0)
		freieKnoten[AnzfreieKnoten++]=z.v;
	    z = z.next;
	}
    }
}
}

class Knoten{
	int Name, anzVorgänger;
	Kante ersterNachfolger;
	Knoten (int i) {
	    Name = i;
	    anzVorgänger = 0; 
	    ersterNachfolger = null;
	}
	public String toString() {return String.valueOf(Name); }
}

class Kante {
	Knoten u,v;
	Kante next;
	Kante(Knoten uu, Knoten vv) {u = uu; v = vv;}
}
